test/Validate.SSA.cpp

```diff
@@ -0,0 +1,699 @@
+// Copyright (c) 2015 The Khronos Group Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+// Validation tests for SSA
+
+#include "UnitSPIRV.h"
+#include "ValidateFixtures.h"
+
+#include <regex>
+
+using std::string;
+using std::regex;
+using std::pair;
+using std::regex_replace;
+namespace {
+
+using Validate = spvtest::ValidateBase<pair<string, bool>>;
+
+TEST_F(Validate, Default) {
+  char str[] = R"(
+     OpMemoryModel Logical GLSL450
+     OpEntryPoint GLCompute %3 ""
+     OpExecutionMode %3 LocalSize 1 1 1
+%1 = OpTypeVoid
+%2 = OpTypeFunction %1
+%3 = OpFunction %1 None %2
+%4 = OpLabel
+     OpReturn
+     OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_SUCCESS);
+}
+
+TEST_F(Validate, IdUndefinedBad) {
+  char str[] = R"(
+     OpMemoryModel Logical GLSL450
+     OpEntryPoint GLCompute %4 ""
+     OpExecutionMode %4 LocalSize 1 1 1
+%2 = OpTypeVoid
+%3 = OpTypeFunction %2
+%4 = OpFunction %2 None %6
+%5 = OpLabel
+     OpReturn
+     OpFunctionEnd
+    )";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, IdRedefinedBad) {
+  char str[] = R"(
+     OpMemoryModel Logical GLSL450
+     OpEntryPoint GLCompute %3 ""
+     OpExecutionMode %3 LocalSize 1 1 1
+%1 = OpTypeVoid
+%2 = OpTypeFunction %1
+%2 = OpFunction %1 None %2
+%4 = OpLabel
+     OpReturn
+     OpFunctionEnd
+)";
+  ValidateInstructions(str, SPV_ERROR_INVALID_ID);
+}
+
+TEST_F(Validate, DominateUsageBad) {
+  char str[] = R"(
+     OpMemoryModel Logical GLSL450
+     OpEntryPoint GLCompute %3 ""
+     OpExecutionMode %3 LocalSize 1 1 1
+%2 = OpTypeFunction %1
+%3 = OpFunction %1 None %2
+%1 = OpTypeVoid
```

Comment:
This can also fail because you've put the void type decl after the start of the function body. So it depends on the order of check done inside the validator.  (Also, you could simplify by avoiding the entry point and function body.)
It would be better if you had just
```
OpMemoryModel Logical GLSL450
%1 = OpTypeFunction %2
%2 = OpTypeVoid
```